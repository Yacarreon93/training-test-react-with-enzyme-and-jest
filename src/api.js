export default {
  addUser: (...args) => {
    console.log("user added", ...args);
  }
};
