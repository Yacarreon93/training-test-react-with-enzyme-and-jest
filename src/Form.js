import React from "react";

import api from "./api";

export default class Form extends React.Component {
  state = {
    name: "",
    email: "",
    phone: "",
    optIn: true
  };

  handleChange = str => e => {
    this.setState({ [str]: e.currentTarget.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    api.addUser(
      this.state.name,
      this.state.email,
      this.state.phone,
      this.state.optIn
    );
  };

  handlePromotionClick = e => {
    this.setState(prevState => ({ optIn: !prevState.optIn }));
  };

  render() {
    return (
      <form data-testid="addUserForm" onSubmit={this.handleSubmit}>
        <h2>INFO</h2>
        <input
          type="text"
          data-testid="name"
          onChange={this.handleChange("name")}
          placeholder="name"
          value={this.state.name}
        />
        <input
          type="text"
          data-testid="email"
          onChange={this.handleChange("email")}
          placeholder="email"
          value={this.state.email}
        />
        <input
          type="text"
          data-testid="phone"
          onChange={this.handleChange("phone")}
          placeholder="phone"
          value={this.state.phone}
        />
        <div>
          <input
            type="checkbox"
            data-testid="optIn"
            onClick={this.handlePromotionClick}
            onChange={this.handlePromotionClick}
            placeholder="optIn"
            checked={this.state.optIn}
          />
          <span data-testid="promotionsP">Receive promotions</span>
        </div>
        <button type="submit" data-testid="submitButton">
          ADD
        </button>
      </form>
    );
  }
}
