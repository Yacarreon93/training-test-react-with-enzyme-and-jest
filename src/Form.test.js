import React from "react";
import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import Form from "./Form";
import api from "./api";

// opted in by default
// actually input their information
// submits the form, calls api
// matches snapshot

configure({ adapter: new Adapter() });

const updateInput = (wrapper, instance, newValue) => {
  const input = wrapper.find(instance);

  input.simulate("change", { currentTarget: { value: newValue } });

  return wrapper.find(instance);
};

describe("<Form />", () => {
  test("receive promotions default is true", () => {
    const wrapper = shallow(<Form />);
    const promotionInput = wrapper.find('[data-testid="optIn"]');

    expect(promotionInput.props().checked).toBe(true);
  });

  test("allows user to fill the form", () => {
    const wrapper = shallow(<Form />);
    const nameInput = updateInput(wrapper, '[data-testid="name"]', "Yasser");
    const emailInput = updateInput(
      wrapper,
      '[data-testid="email"]',
      "test@test.com"
    );
    const phoneInput = updateInput(
      wrapper,
      '[data-testid="phone"]',
      "1234567890"
    );
    wrapper.find('[type="checkbox"]').simulate("click");

    expect(nameInput.props().value).toBe("Yasser");
    expect(emailInput.props().value).toBe("test@test.com");
    expect(phoneInput.props().value).toBe("1234567890");
    expect(wrapper.find('[type="checkbox"]').props().checked).toBe(false);
  });

  test("submits the form", () => {
    jest
      .spyOn(api, "addUser")
      .mockImplementation(() => Promise.resolve({ data: "new user added" }));

    const wrapper = shallow(<Form />);
    updateInput(wrapper, '[data-testid="name"]', "Yasser");
    updateInput(wrapper, '[data-testid="email"]', "test@test.com");
    updateInput(wrapper, '[data-testid="phone"]', "1234567890");
    wrapper.find('[data-testid="addUserForm"]').simulate("submit", { preventDefault: () => {} });

    expect(api.addUser).toHaveBeenCalledWith(
      "Yasser",
      "test@test.com",
      "1234567890",
      true
    );
  });

  test("matches snapshot", () => {
    const wrapper = shallow(<Form />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
