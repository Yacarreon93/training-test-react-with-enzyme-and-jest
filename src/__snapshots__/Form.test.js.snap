// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches snapshot 1`] = `
<form
  data-testid="addUserForm"
  onSubmit={[Function]}
>
  <h2>
    INFO
  </h2>
  <input
    data-testid="name"
    onChange={[Function]}
    placeholder="name"
    type="text"
    value=""
  />
  <input
    data-testid="email"
    onChange={[Function]}
    placeholder="email"
    type="text"
    value=""
  />
  <input
    data-testid="phone"
    onChange={[Function]}
    placeholder="phone"
    type="text"
    value=""
  />
  <div>
    <input
      checked={true}
      data-testid="optIn"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="optIn"
      type="checkbox"
    />
    <span
      data-testid="promotionsP"
    >
      Receive promotions
    </span>
  </div>
  <button
    data-testid="submitButton"
    type="submit"
  >
    ADD
  </button>
</form>
`;
